python/cuda_parallel/cuda/parallel/experimental/algorithms/reduce.py

```diff
@@ -46,6 +46,30 @@ def _dtype_validation(dt1, dt2):
         raise TypeError(f"dtype mismatch: __init__={dt1}, __call__={dt2}")
 
 
+def _validate_and_get_stream(stream) -> Optional[int]:
```

Comment:
In terms of where this function should live, here are a couple of suggestions:

- Move it to `_utils/stream.py`
- Rename `_utils/cai.py` to `_utils/protocols.py` and move it there (this module could be general utilities for working with protocol objects like `__cuda_array_interface__` and `__cuda_stream__`)